pkg/server/plugin/datastore/sql/sql.go

```diff
@@ -1124,149 +1122,761 @@ func createRegistrationEntry(tx *gorm.DB,
+UNION
+
+SELECT
+	F.registered_entry_id, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, B.trust_domain, NULL, NULL
+FROM
+	bundles B
+INNER JOIN
+	federated_registration_entries F
+ON
+	B.id = F.bundle_id
+WHERE
+	F.registered_entry_id IN (SELECT id FROM listing)
+
+UNION
+
+SELECT
+	registered_entry_id, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, id, value
+FROM
+	dns_names
+WHERE registered_entry_id IN (SELECT id FROM listing)
+
+UNION
+
+SELECT
+	registered_entry_id, NULL, NULL, NULL, NULL, NULL, NULL, NULL, id, type, value, NULL, NULL, NULL
+FROM
+	selectors
+WHERE registered_entry_id IN (SELECT id FROM listing)
+
+ORDER BY selector_id, dns_name_id
+;`
+	return query, []interface{}{req.EntryId}, nil
+}
+
+func buildFetchRegistrationEntryQueryMySQL(req *datastore.FetchRegistrationEntryRequest) (string, []interface{}, error) {
+	const query = `
+SELECT
+	E.id AS e_id,
+	E.entry_id AS entry_id,
+	E.spiffe_id,
+	E.parent_id,
+	E.ttl AS reg_ttl,
+	E.admin,
+	E.downstream,
+	E.expiry,
+	S.id AS selector_id,
+	S.type AS selector_type,
+	S.value AS selector_value,
+	B.trust_domain,
+	D.id AS dns_name_id,
+	D.value AS dns_name
+FROM
+	registered_entries E
+LEFT JOIN
+	(SELECT 1 AS joinItem UNION SELECT 2 UNION SELECT 3) AS joinItems ON TRUE
+LEFT JOIN
+	selectors S ON joinItem=1 AND E.id=S.registered_entry_id
+LEFT JOIN
+	dns_names D ON joinItem=2 AND E.id=D.registered_entry_id
+LEFT JOIN
+	(federated_registration_entries F INNER JOIN bundles B ON F.bundle_id=B.id) ON joinItem=3 AND E.id=F.registered_entry_id
+WHERE E.entry_id = ?
+ORDER BY selector_id, dns_name_id
+;`
+	return query, []interface{}{req.EntryId}, nil
+}
+
+func listRegistrationEntries(ctx context.Context, dbType string, db *sql.DB, req *datastore.ListRegistrationEntriesRequest) (*datastore.ListRegistrationEntriesResponse, error) {
+	if req.Pagination != nil && req.Pagination.PageSize == 0 {
+		return nil, status.Error(codes.InvalidArgument, "cannot paginate with pagesize = 0")
+	}
+	if req.BySelectors != nil && len(req.BySelectors.Selectors) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "cannot list by empty selector set")
+	}
+
+	for {
+		resp, err := listRegistrationEntriesOnce(ctx, dbType, db, req)
+		if err != nil {
+			return nil, err
+		}
+
+		if req.BySelectors == nil || len(resp.Entries) == 0 {
+			return resp, nil
+		}
+
+		// Exact/subset matching requires filtering out all registration
+		// entries returned by the query whose selectors are not fully
+		// represented in the request selectors.
+		// TODO: we could probably push this into SQL with some fancy "group
+		// by" count filtering.
+		type selectorKey struct {
+			Type  string
+			Value string
+		}
+		set := make(map[selectorKey]struct{}, len(req.BySelectors.Selectors))
+		for _, s := range req.BySelectors.Selectors {
+			set[selectorKey{Type: s.Type, Value: s.Value}] = struct{}{}
+		}
+		filtered := make([]*common.RegistrationEntry, 0, len(resp.Entries))
+	entryLoop:
```

Comment:
I'm not a fan of the loop label, I'd prefer we move stuff into other functions to avoid this and double check the compiler inlines them.